### PR TITLE
Wetonomy Members

### PR DESCRIPTION
**Pull request by IoanStoianov**

### CODE REVIEW

File: src/Members/Members/MembersRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+			: base(address, registry, permissionManager)
+		{
+			this.membersAddresses = addressesToMembers ?? new HashSet<Address>();
+			this.registry = registry;
+			this.ConfigurePermissionManager(permissionManager);
+		}
+
+		public HashSet<Member> GetAllMembers()
+		{
+			var members = new HashSet<Member>();
```
Comment (ivanasen) [inline src/Members/Members/MembersRegistry.cs]: Can't we store a HashSet of the Members contracts and not only their addresses so we don't have to create a new set every time

File: src/Members.Tests/AddressGroupRegistryTest.cs [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+		{
+			var group = new AddressGroup(this.addressFactory.Create(), this.contractRegistry, this.permissionManager.Address);
+			var addMemberAction = new RegisterGroupAction(string.Empty, this.addressGroupRegistry.Address, group);
+			this.permissionManager.ExecuteAction(addMemberAction);
+			var mem = this.addressGroupRegistry.GetGroup(group.Address);
+			Assert.Equal(group, mem);
+		}
+
+		[Fact]
+		public void AddMember_WhenAddresDuplicate_AssertEqual()
```
Comment (zh-m) [inline src/Members.Tests/AddressGroupRegistryTest.cs]: AssertEqual is expected behavior?

File: src/Members.Tests/AddressGroupRegistryTest.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+			var addMemberAction = new RegisterGroupAction(string.Empty, this.addressGroupRegistry.Address, group);
+			this.permissionManager.ExecuteAction(addMemberAction);
+			Assert.Single(this.addressGroupRegistry.GetAllGroups());
+			addMemberAction = new RegisterGroupAction(string.Empty, this.addressGroupRegistry.Address, group);
+			this.permissionManager.ExecuteAction(addMemberAction);
+			Assert.Single(this.addressGroupRegistry.GetAllGroups());
+		}
+
+		[Fact]
+		public void RemoveMember_ReturnNull()
```
Comment (zh-m) [inline src/Members.Tests/AddressGroupRegistryTest.cs]: RemoveMember - returns null ? (probably not the correct name for a test method)

File: src/Members/AddressGroups/Actions/GroupActions/UnregisterGroupMemberAction.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+﻿using ContractsCore;
+using ContractsCore.Actions;
+
+namespace Members.AddressGroups.Actions
+{
+	public class UnregisterGroupMemberAction : Action
+	{
+		public Address MemberAddress;
```
Comment (zh-m) [inline src/Members/AddressGroups/Actions/GroupActions/UnregisterGroupMemberAction.cs]: fields are named will small letter, properties - with capital.

File: src/Members/AddressGroups/AddressGroup.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+				default:
+					return false;
+			}
+		}
+
+		protected virtual bool HandleAddMemberAction(RegisterGroupMemberAction action)
+		{
+			this.members.Add(action.MemberAddress);
+			return true;
```
Comment (zh-m) [inline src/Members/AddressGroups/AddressGroup.cs]: Why should we return bool in a method called Handle..?

File: src/Members/AddressGroups/AddressGroupRegistry.cs [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+			try
+			{
+				this.registry.GetContract(group.Address);
+			}
+			catch (KeyNotFoundException)
+			{
+				this.registry.RegisterContract(group);
+			}
+
+			this.groupAddresses.Add(group.Address);
```
Comment (zh-m) [inline src/Members/AddressGroups/AddressGroupRegistry.cs]: What if there's an exception - should we add the address?

File: src/Members/AddressGroups/AddressGroupRegistry.cs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+		{
+			return this.groupAddresses.Contains(address) ? (this.registry.GetContract(address) as AddressGroup) : null;
+		}
+
+		public HashSet<AddressGroup> GetAllGroups()
+		{
+			var members = new HashSet<AddressGroup>();
+			foreach (var address in groupAddresses)
+			{
+				members.Add(this.registry.GetContract(address) as AddressGroup);
```
Comment (zh-m) [inline src/Members/AddressGroups/AddressGroupRegistry.cs]: Should we initialize them every time in a method GetAllGroups or probably cache?

**Pull request merged**